File: 07-ios-swift/SilkVoice/CallViewModel.swift
// CallViewModel.swift — bridges LiveKit's RoomDelegate into SwiftUI state.
//
// UI state machine:
//   idle → connecting → listening ⇄ thinking ⇄ speaking → (interrupting) → … → idle
//
// `speaking` is red; `interrupting` (the caller talks over the agent) is
// green. State comes from the agent's `lk.agent.state` attribute when it
// publishes one, and from isSpeaking updates otherwise.

import AVFoundation
import Foundation
import LiveKit

enum AgentUIState: String {
    case idle, connecting, listening, thinking, speaking, interrupting, error

    var label: String {
        switch self {
        case .idle: return "Idle"
        case .connecting: return "Connecting…"
        case .listening: return "Listening"
        case .thinking: return "Thinking"
        case .speaking: return "Agent speaking"
        case .interrupting: return "You interrupted — go ahead"
        case .error: return "Error"
        }
    }
}

private let agentStateAttribute = "lk.agent.state"

@MainActor
final class CallViewModel: ObservableObject {
    @Published private(set) var state: AgentUIState = .idle
    @Published private(set) var errorMessage: String?
    @Published private(set) var callId: String?
    @Published private(set) var outputRoute: String = AudioRouting.currentOutputName()
    @Published var speakerphone = true {
        didSet { AudioRouting.setSpeakerphone(speakerphone) }
    }

    let room = Room()

    private var agentAttrState: String?
    private var agentSpeaking = false
    private var userSpeaking = false
    private var routeObserver: NSObjectProtocol?

    var inCall: Bool { room.connectionState == .connected || room.connectionState == .reconnecting }

    init() {
        room.add(delegate: self)
        // Bluetooth headphones connect, CarPlay takes over, user pulls out the
        // wired earphones… keep the label honest.
        routeObserver = NotificationCenter.default.addObserver(
            forName: AVAudioSession.routeChangeNotification, object: nil, queue: .main
        ) { [weak self] _ in
            Task { @MainActor in self?.outputRoute = AudioRouting.currentOutputName() }
        }
    }

    func start() async {
        errorMessage = nil
        state = .connecting
        do {
            // 1. Microphone permission first: /v1/webcall bills the moment it returns.
            guard await AudioRouting.requestMicrophonePermission() else {
                throw SilkAPIError(status: 0, code: "mic_denied", message: "Microphone permission denied.")
            }
            // 2. Credentials from your backend.
            let creds = try await SilkAPI.startCall()
            callId = creds.callId

            // 3. Join and publish the microphone. The agent greets as soon as it sees us.
            try await room.connect(url: creds.host, token: creds.token)
            try await room.localParticipant.setMicrophone(enabled: true)
            AudioRouting.setSpeakerphone(speakerphone)

            // 4. The agent may already be in the room with a state attribute.
            for participant in room.remoteParticipants.values {
                if let s = participant.attributes[agentStateAttribute] { agentAttrState = s }
            }
            recompute()
        } catch {
            errorMessage = error.localizedDescription
            state = .error
            await room.disconnect()
        }
    }

    func hangUp() async {
        await room.disconnect() // → didUpdateConnectionState(.disconnected) → idle; billing stops
    }

    private func recompute() {
        var s: AgentUIState
        switch agentAttrState {
        case "speaking": s = .speaking
        case "thinking": s = .thinking
        case "listening": s = .listening
        case "initializing": s = .connecting
        default: s = agentSpeaking ? .speaking : .listening // no attribute → who is making noise
        }
        // Barge-in: the caller speaks while the agent still counts as speaking.
        if s == .speaking, userSpeaking { s = .interrupting }
        state = s
    }

    private func reset() {
        agentAttrState = nil
        agentSpeaking = false
        userSpeaking = false
    }
}

// MARK: - RoomDelegate → SwiftUI

extension CallViewModel: RoomDelegate {
    nonisolated func room(_ room: Room, participant: Participant, didUpdateIsSpeaking isSpeaking: Bool) {
        Task { @MainActor in
            if participant is LocalParticipant { userSpeaking = isSpeaking } else { agentSpeaking = isSpeaking }
            recompute()
        }
    }

    nonisolated func room(_ room: Room, participant: Participant, didUpdateAttributes attributes: [String: String]) {
        guard !(participant is LocalParticipant), let s = attributes[agentStateAttribute] else { return }
        Task { @MainActor in
            agentAttrState = s
            recompute()
        }
    }

    nonisolated func room(_ room: Room, didUpdateConnectionState connectionState: ConnectionState, from oldConnectionState: ConnectionState) {
        Task { @MainActor in
            switch connectionState {
            case .disconnected:
                reset()
                if state != .error { state = .idle }
            case .connecting, .reconnecting:
                state = .connecting
            case .connected:
                recompute()
            }
        }
    }

    nonisolated func room(_ room: Room, participant: RemoteParticipant, didSubscribeTrack publication: RemoteTrackPublication) {
        // Remote audio plays through the active AVAudioSession route automatically.
    }
}

File: 08-android-kotlin/app/src/main/java/ai/rumik/silkvoice/CallViewModel.kt
// CallViewModel.kt — bridges LiveKit's RoomEvent flow into Compose state.
//
// UI state machine:
//   IDLE → CONNECTING → LISTENING ⇄ THINKING ⇄ SPEAKING → (INTERRUPTING) → … → IDLE
//
// SPEAKING is red; INTERRUPTING (the caller talks over the agent) is green.
// State comes from the agent's `lk.agent.state` attribute when it publishes
// one, and from ActiveSpeakersChanged otherwise.
package ai.rumik.silkvoice

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.twilio.audioswitch.AudioDevice
import io.livekit.android.AudioOptions
import io.livekit.android.AudioType
import io.livekit.android.LiveKit
import io.livekit.android.LiveKitOverrides
import io.livekit.android.RoomOptions
import io.livekit.android.events.RoomEvent
import io.livekit.android.events.collect
import io.livekit.android.room.Room
import io.livekit.android.room.participant.LocalParticipant
import io.livekit.android.room.participant.RemoteParticipant
import io.livekit.android.room.track.LocalAudioTrackOptions
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

enum class AgentUiState(val label: String) {
    IDLE("Idle"),
    CONNECTING("Connecting…"),
    LISTENING("Listening"),
    THINKING("Thinking"),
    SPEAKING("Agent speaking"),
    INTERRUPTING("You interrupted — go ahead"),
    ERROR("Error"),
}

data class CallUi(
    val state: AgentUiState = AgentUiState.IDLE,
    val error: String? = null,
    val callId: String? = null,
    val inCall: Boolean = false,
    val outputDevice: String = "—",
    val availableDevices: List<AudioDevice> = emptyList(),
)

private const val AGENT_STATE_ATTR = "lk.agent.state"

class CallViewModel(app: Application) : AndroidViewModel(app) {

    private val _ui = MutableStateFlow(CallUi())
    val ui: StateFlow<CallUi> = _ui

    // One Room for the lifetime of the ViewModel; connect/disconnect per call.
    // AudioType.CallAudioType puts the AudioManager into MODE_IN_COMMUNICATION with
    // USAGE_VOICE_COMMUNICATION: that is what enables the hardware echo canceller
    // so the agent does not hear itself out of the phone speaker.
    private val room: Room = LiveKit.create(
        appContext = app,
        options = RoomOptions(
            adaptiveStream = true,
            dynacast = true,
            audioTrackCaptureDefaults = LocalAudioTrackOptions(
                echoCancellation = true,
                noiseSuppression = true,
                autoGainControl = true,
            ),
        ),
        overrides = LiveKitOverrides(
            audioOptions = AudioOptions(audioOutputType = AudioType.CallAudioType()),
        ),
    )
    val routing = AudioRouting(room)

    private var eventsJob: Job? = null
    private var agentAttrState: String? = null
    private var agentSpeaking = false
    private var userSpeaking = false

    fun start() {
        if (_ui.value.inCall) return
        _ui.update { it.copy(error = null, state = AgentUiState.CONNECTING) }
        viewModelScope.launch {
            try {
                // 1. Credentials from your backend (RECORD_AUDIO was requested by the Activity first —
                //    /v1/webcall bills the moment it returns, so never start a call you cannot join).
                val creds = SilkApi.startCall()

                // 2. Subscribe to room events before connecting so nothing is missed.
                eventsJob = launch { room.events.collect { onRoomEvent(it) } }

                // 3. Join and publish the microphone. The agent greets as soon as it sees us.
                room.connect(creds.host, creds.token)
                room.localParticipant.setMicrophoneEnabled(true)

                // 4. The agent may already be in the room with a state attribute.
                room.remoteParticipants.values.forEach { p ->
                    p.attributes[AGENT_STATE_ATTR]?.let { agentAttrState = it }
                }
                _ui.update { it.copy(inCall = true, callId = creds.callId) }
                refreshRouting()
                recompute()
            } catch (e: Exception) {
                eventsJob?.cancel()
                _ui.update { it.copy(state = AgentUiState.ERROR, error = e.toString(), inCall = false) }
            }
        }
    }

    fun hangUp() {
        viewModelScope.launch { room.disconnect() } // → RoomEvent.Disconnected → IDLE; billing stops
    }

    fun selectDevice(device: AudioDevice) {
        routing.select(device)
        refreshRouting()
    }

    private fun onRoomEvent(event: RoomEvent) {
        when (event) {
            is RoomEvent.ActiveSpeakersChanged -> {
                agentSpeaking = event.speakers.any { it is RemoteParticipant }
                userSpeaking = event.speakers.any { it is LocalParticipant }
                recompute()
            }
            is RoomEvent.ParticipantAttributesChanged -> {
                if (event.participant is RemoteParticipant) {
                    event.changedAttributes[AGENT_STATE_ATTR]?.let { agentAttrState = it; recompute() }
                }
            }
            is RoomEvent.TrackSubscribed -> {
                // Remote audio plays automatically; nothing to attach on Android.
            }
            is RoomEvent.Disconnected -> {
                eventsJob?.cancel()
                agentAttrState = null; agentSpeaking = false; userSpeaking = false
                _ui.update { it.copy(state = AgentUiState.IDLE, inCall = false) }
            }
            else -> Unit
        }
    }

    private fun recompute() {
        var s = when (agentAttrState) {
            "speaking" -> AgentUiState.SPEAKING
            "thinking" -> AgentUiState.THINKING
            "listening" -> AgentUiState.LISTENING
            "initializing" -> AgentUiState.CONNECTING
            else -> if (agentSpeaking) AgentUiState.SPEAKING else AgentUiState.LISTENING // no attribute → who is making noise
        }
        // Barge-in: the caller speaks while the agent still counts as speaking.
        if (s == AgentUiState.SPEAKING && userSpeaking) s = AgentUiState.INTERRUPTING
        _ui.update { it.copy(state = s) }
    }

    private fun refreshRouting() {
        _ui.update {
            it.copy(
                outputDevice = AudioRouting.label(routing.selectedDevice),
                availableDevices = routing.availableDevices,
            )
        }
    }

    override fun onCleared() {
        viewModelScope.launch { room.disconnect() }
        room.release()
        super.onCleared()
    }
}
